server/datastore/mysql/scripts.go

```diff
@@ -1795,39 +1796,70 @@ func (ds *Datastore) getOrGenerateScriptContentsID(ctx context.Context, contents
 	return scriptContentsID, nil
 }
 
-func (ds *Datastore) BatchExecuteScript(ctx context.Context, userID *uint, scriptID uint, hostIDs []uint) (string, error) {
+func teamIDEq(teamID1, teamID2 *uint) bool {
+	sameTeamNoTeam := teamID1 == nil && teamID2 == nil
+	sameTeamNumber := teamID1 != nil && teamID2 != nil && *teamID1 == *teamID2
+	return sameTeamNoTeam || sameTeamNumber
+}
+
+func (ds *Datastore) batchExecuteScript(ctx context.Context, userID *uint, scriptID uint, hostIDs []uint, batchExecID string) error {
 	script, err := ds.Script(ctx, scriptID)
 	if err != nil {
-		return "", fleet.NewInvalidArgumentError("script_id", err.Error())
+		return fleet.NewInvalidArgumentError("script_id", err.Error())
 	}
 
+	invalidHostIDPlatform := "batch-invalid-hostid"
+	invalidHostTeamPlatform := "batch-invalid-teamid"
+
 	// We need full host info to check if hosts are able to run scripts, see svc.RunHostScript
 	fullHosts := make([]*fleet.Host, 0, len(hostIDs))
 
+	// The execution results to be stored in the database
+	executions := make([]fleet.BatchExecutionHost, 0, len(fullHosts))
+
 	// Check that all hosts exist before attempting to process them
 	for _, hostID := range hostIDs {
 		host, err := ds.Host(ctx, hostID)
 		if err != nil {
-			return "", fmt.Errorf("unable to load host information for %d: %w", hostID, err)
+			fullHosts = append(fullHosts, &fleet.Host{
+				ID:       hostID,
+				Platform: invalidHostIDPlatform,
+			})
+			continue
 		}
 
 		// All hosts must be on the same team as the script
-		sameTeamNoTeam := host.TeamID == nil && script.TeamID == nil
-		sameTeamNumber := host.TeamID != nil && script.TeamID != nil && *host.TeamID == *script.TeamID
-		sameTeam := sameTeamNoTeam || sameTeamNumber
-		if !sameTeam {
-			return "", ctxerr.Errorf(ctx, "all hosts must be on the same team as the script")
+		if !teamIDEq(host.TeamID, script.TeamID) {
+			fullHosts = append(fullHosts, &fleet.Host{
+				ID:       hostID,
+				Platform: invalidHostTeamPlatform,
+			})
+			continue
 		}
 
 		fullHosts = append(fullHosts, host)
 	}
 
-	executions := make([]fleet.BatchExecutionHost, 0, len(fullHosts))
-
-	batchExecID := ""
-
 	if err := ds.withRetryTxx(ctx, func(tx sqlx.ExtContext) error {
 		for _, host := range fullHosts {
+			// Host doesn't exist anymore
+			if host.Platform == invalidHostIDPlatform {
+				executions = append(executions, fleet.BatchExecutionHost{
+					HostID: host.ID,
+					Error:  &fleet.BatchExecuteInvalidHost,
+				})
+				continue
+			}
+
+			// Host isn't on the right team anymore, can't execure script
+			if host.Platform == invalidHostTeamPlatform {
+				executions = append(executions, fleet.BatchExecutionHost{
+					HostID: host.ID,
+					Error:  &fleet.BatchExecuteIncompatibleTeam,
+				})
+				continue
+			}
```

Comment:
I think we decided to still execute the script in this case.  @rachaelshaw did we agree that if a host switches teams after being scheduled to run a script, we should still run it when the schedule comes due?  Because the user selected those hosts specifically?

Reply:
@sgress454 actually, not sure we fully specced out this edge case; we specified that the targeted hosts don't change, but didn't clarify their behavior when they try to run another team's script.

What's the current behavior when a script is queued up for an offline host, then the host is transferred before they come online & the script runs? We should probably mimic that.

Reply:
From the looks of the code, it'll still run the script, but I can test.